Implemented SIMD trunc instructions.

danlliu opened the pull request:
#### 46168caf6713a0f0287f2fa5b04861e586750c6d
<pre>
Implemented SIMD trunc instructions.
<a href="https://bugs.webkit.org/show_bug.cgi?id=257923">https://bugs.webkit.org/show_bug.cgi?id=257923</a>
rdar://problem/110554488

Reviewed by Justin Michaud.

Implemented and added tests for relaxed_trunc_f32x4_s, relaxed_trunc_f32x4_u, relaxed_trunc_f64x2_s_zero, and relaxed_trunc_64x2_u_zero.

* JSTests/wasm/stress/simd-const-relaxed-f32-trunc.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.sz.i32.result.i32.i32.const.9.drop.v128.const.f32x4.1.5.2.4.3.3.4.2.i32x4.relaxed_trunc_f32x4_s.return.i32x4.extract_lane.1.func.export.string_appeared_here.param.sz.i32.result.i32.v128.const.f32x4.1.4.2.3.3.2.4.1.i32x4.relaxed_trunc_f32x4_s.return.i32x4.extract_lane.1.func.export.string_appeared_here.param.sz.i32.result.i32.v128.const.f32x4.1.5.2.4.3.3.4.2.i32x4.relaxed_trunc_f32x4_u.return.i32x4.extract_lane.1.func.export.string_appeared_here.param.sz.i32.result.i32.v128.const.f32x4.14.0.1.39.16.5.1.0.i32x4.relaxed_trunc_f32x4_u.return.i32x4.extract_lane.2.async test):
* JSTests/wasm/stress/simd-const-relaxed-f64-trunc.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.sz.i32.result.i32.i32.const.9.drop.v128.const.f64x2.1.5.1.5.i32x4.relaxed_trunc_f64x2_s_zero.return.i32x4.extract_lane.1.func.export.string_appeared_here.param.sz.i32.result.i32.v128.const.f64x2.1.4.2.3.i32x4.relaxed_trunc_f64x2_s_zero.return.i32x4.extract_lane.2.func.export.string_appeared_here.param.sz.i32.result.i32.v128.const.f64x2.1.5.2.4.i32x4.relaxed_trunc_f64x2_u_zero.return.i32x4.extract_lane.1.func.export.string_appeared_here.param.sz.i32.result.i32.v128.const.f64x2.14.0.1.39.i32x4.relaxed_trunc_f64x2_u_zero.return.i32x4.extract_lane.2.async test):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
(JSC::B3::ValueKey::materialize const):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addSIMDV_V):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::addSIMDV_V):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::simd):
* Source/JavaScriptCore/wasm/WasmSIMDOpcodes.h:
(JSC::dumpSIMDLaneOperation):
(JSC::isRelaxedSIMDOperation):

Canonical link: <a href="https://commits.webkit.org/265136@main">https://commits.webkit.org/265136@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f024ec7169ac07ed928d75b2654184c5a94f1f59

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9928 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10175 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10425 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11580 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9625 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12161 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10127 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12565 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10082 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10877 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8412 "2 api tests failed or timed out") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11963 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8184 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9003 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16341 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8416 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9282 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9152 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12423 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/9428 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9654 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7843 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/10071 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8813 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2576 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2384 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13039 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/10341 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9428 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2561 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator64.cpp:
```diff
@@ -272,6 +272,8 @@ class AirIRGenerator64 : public AirIRGeneratorBase<AirIRGenerator64, TypedTmp> {
         AIR_OP_CASE(Not)
         AIR_OP_CASE(Neg)
 
+        else if (op == SIMDLaneOperation::RelaxedTruncSat) airOp = B3::Air::VectorTruncSat;
```
You can skip this file, we don't use this tier

danlliu replied:
Commenting out this line leads to the `wasm-air` tests failing, since OMG can't translate `RelaxedTruncSat` into any Air opcode.

justinmichaud reviewed (approved):
r=me

To fix the test failures, you should skip this test on intel. That is, add a new //@ skip if line to the test to skip on non-arm64 platforms.

danlliu committed:
Implemented SIMD trunc instructions.
https://bugs.webkit.org/show_bug.cgi?id=257923
rdar://problem/110554488

Reviewed by Justin Michaud.

Implemented and added tests for relaxed_trunc_f32x4_s, relaxed_trunc_f32x4_u, relaxed_trunc_f64x2_s_zero, and relaxed_trunc_64x2_u_zero.

* JSTests/wasm/stress/simd-const-relaxed-f32-trunc.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.sz.i32.result.i32.i32.const.9.drop.v128.const.f32x4.1.5.2.4.3.3.4.2.i32x4.relaxed_trunc_f32x4_s.return.i32x4.extract_lane.1.func.export.string_appeared_here.param.sz.i32.result.i32.v128.const.f32x4.1.4.2.3.3.2.4.1.i32x4.relaxed_trunc_f32x4_s.return.i32x4.extract_lane.1.func.export.string_appeared_here.param.sz.i32.result.i32.v128.const.f32x4.1.5.2.4.3.3.4.2.i32x4.relaxed_trunc_f32x4_u.return.i32x4.extract_lane.1.func.export.string_appeared_here.param.sz.i32.result.i32.v128.const.f32x4.14.0.1.39.16.5.1.0.i32x4.relaxed_trunc_f32x4_u.return.i32x4.extract_lane.2.async test):
* JSTests/wasm/stress/simd-const-relaxed-f64-trunc.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.sz.i32.result.i32.i32.const.9.drop.v128.const.f64x2.1.5.1.5.i32x4.relaxed_trunc_f64x2_s_zero.return.i32x4.extract_lane.1.func.export.string_appeared_here.param.sz.i32.result.i32.v128.const.f64x2.1.4.2.3.i32x4.relaxed_trunc_f64x2_s_zero.return.i32x4.extract_lane.2.func.export.string_appeared_here.param.sz.i32.result.i32.v128.const.f64x2.1.5.2.4.i32x4.relaxed_trunc_f64x2_u_zero.return.i32x4.extract_lane.1.func.export.string_appeared_here.param.sz.i32.result.i32.v128.const.f64x2.14.0.1.39.i32x4.relaxed_trunc_f64x2_u_zero.return.i32x4.extract_lane.2.async test):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
(JSC::B3::ValueKey::materialize const):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addSIMDV_V):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::addSIMDV_V):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::simd):
* Source/JavaScriptCore/wasm/WasmSIMDOpcodes.h:
(JSC::dumpSIMDLaneOperation):
(JSC::isRelaxedSIMDOperation):

Canonical link: https://commits.webkit.org/265136@main
